### PR TITLE
Docker configuration 418

### LINKED ISSUE #418: Create a docker image of the environment for convenience
Creating a docker image of our environment would make setting up the env a painless process (just launch the image). This would allow for more rigorous testing and bold modifications because people wouldn't be afraid to nuke their install 🙂 

Also potentially travis would be able to launch the docker image, which would greatly reduce the time per build since currently about 2.5 mins is dedicated to installing apt packages.

### PR DESCRIPTION
**Pull request by MewtR**
# Assignee Section

## Description
This PR dockerizes the project, allowing users to run it regardless of their environment.
Using docker-compose, three services (group of containers based on the same docker image) are created:
- One running roscore called `ros-master`
- One running rosbridge called `rosbridge`
- One running the actual project called `prototype`

### Steps for Testing

1. Follow the `README` instructions in the `docker` directory to run the project.
2. Once everything is up and running, make sure you see the GUI at http://localhost:5000
3. Open a terminal and run `docker exec -it docker_rosbridge_1 bash`. This gives you a shell inside the container running rosbridge
4. Still inside the container, run `rostopic pub -1 /battery_voltage mcu_control/Voltage 15.0`
5. You should see the voltage displayed on the GUI

closes #418 (doesn't include the travis part)

# Reviewer Section

Aside from local testing and the General Integration Test it is implied that static analysis should be included in the verification process.

- [ ] Local Test Performed Successfully
- [ ] [General Integration Test](https://docs.google.com/document/d/1ug0CpA1cIzURP8DDFSvCt2CEJJSwJ6Ta6B1LG_hYk6I/edit) Performed Successfully

For Pull Requests that do not include code changes, it is not required to perform the tests above.

### CODE REVIEW

**Comment by MartensCedric** [comment]
> #418

It's already part of the description :sunglasses: so it will close the issue when the PR closes

File: Dockerfile
Comment (MartensCedric) [inline Dockerfile:9]: I suggest to use /home/spaceuser/Programming/robotics-prototype
We have a file with handy dandy aliases like `robotics` that `cd`s to the robotics directory, or `cm` that `cd`s into the rospackages and runs catkin_make, and a lot more aliases, but they currently assume that you have a `Programming` folder like this `~/Programming/robotics-prototype`

**Review by MartensCedric — CHANGES_REQUESTED** [review]

**Review by AFknUsername — COMMENTED** [review]
Managed to launch it on my machine and websocket connected.
Will recheck when @MartensCedric 's comment is implemented

**Comment by RealWilliamWells** [comment]
Screenshot from 2020-10-03 13-09-52

The published values don't show up for me in the GUI for some reason, but I think this is some weird issue with docker on my end. I have noticed this tho, I think net-tools needs to be installed in the docker in order to have access to the ip-config command.

File: Dockerfile
Comment (MewtR) [reply Dockerfile]: So this is done. See fff03f2

**Comment by MewtR** [comment]
@RealWilliamWells 
I'm actually facing the exact same issue as you. I tried installing net-tools but that doesn't seem to fix the issue. I get several log messages saying that rosbridge can't subscribe to certain topics.

**Comment by MartensCedric** [comment]
So I pulled and did `sudo docker-compose up` it recreated the container I believe but I don't see the `Programming` directory. Is there additional steps I must do

File: Dockerfile
Comment (MartensCedric) [inline Dockerfile:18]: Shouldn't you do something like COPY . ~/Programming/

**Comment by MartensCedric** [comment]
Two things you should know @MewtR 

On our normal install we make sure that the ROS things are properly sourced from the ~/.bashrc we also have some handy dandy aliases

https://github.com/space-concordia-robotics/robotics-prototype/blob/master/EnvironmentSetup.sh

You might wanna at least add the source setup.bash so we dont have to do it everytime

**Comment by RealWilliamWells** [comment]
image
I get this error now when trying to run `rostopic pub -1 /battery_voltage mcu_control/Voltage 15.0` inside the docker running rosbridge.

File: Dockerfile
Comment (MewtR) [reply Dockerfile:18]: Further up there is a line that sets the working directory to /home/spaceuser/Programming/robotics-prototype (`WORKDIR /home/spaceuser/Programming/robotics-prototype`). Like I said earlier just make sure you `sudo docker-compose up --build` to take changes into account

**Comment by MewtR** [comment]
@RealWilliamWells 
Please try it from the rosbridge container with `docker exec -it docker_rosbridge_1 bash`. I didn't build the rospackages in the prototype container because it didn't seem necessary.

@MartensCedric  
So I added the automatic sourcing you're referring to. See 2170505

**Comment by RealWilliamWells** [comment]
> @RealWilliamWells
> Please try it from the rosbridge container with `docker exec -it docker_rosbridge_1 bash`. I didn't build the rospackages in the prototype container because it didn't seem necessary.
> 
> @MartensCedric
> So I added the automatic sourcing you're referring to. See [2170505](https://github.com/space-concordia-robotics/robotics-prototype/commit/21705052b52e38559f661c66e11f856448922034)

I did try it in the rosbridge container with `docker exec -it docker_rosbridge_1 bash`. But I forgot to mention something else, for some reason to be able to run this command I have to manually source `/opt/ros/melodic/setup.bash`. From what I can tell it seems as if things aren't being sourced properly when starting up a terminal in the docker. Although, I believe that the issue with publishing voltage values after sourcing `/opt/ros/melodic/setup.bash` may just be an issue on my end, given that this seemed to have worked for most other people.

**Comment by MewtR** [comment]
@RealWilliamWells 
Actually the manual sourcing was how it was done initially. I only added the auto sourcing like 2 days ago. So just to be sure, you pulled the latest changes and still had to manually source `/opt/ros/melodic/setup.bash`? Also did you make sure you ran `docker-compose up --build`?

**Comment by RealWilliamWells** [comment]
@MewtR 
I tried re-pulling and ran `docker-compose up --build`, but I get these errors when running the command with `--build`:
image

image

Attempting to run `rostopic` within the container gives the same result as before.

**Comment by MewtR** [comment]
@RealWilliamWells 
Not sure why you get that error. My best guess right now is that your docker might be running out of space.
Can you try:
* `docker rm $(docker container ls -aq)`
* `docker image prune -a` <- deletes all images not associated with  a container
* And then just try the `docker-compose up --build` again? Problem of course is that is has to re-download everything.

File: docker/rosbridge/Dockerfile
Comment (RealWilliamWells) [inline docker/rosbridge/Dockerfile:17]: Adding "RUN apt-get update -y && apt-get upgrade -y" here fixed my errors with building the docker images.

**Review by RealWilliamWells — COMMENTED** [review]
@MewtR 
I managed to fix it and everything works as intended! I tried what you suggested and attempted to rebuild the dockers but to no avail. What ended up working was running "apt-get update -y && apt-get upgrade -y" before building the rospackages in the dockerfile. I made a comment on that line in the file to show where I put it in order to fix the errors I was getting. Now that I'm able to properly build the rospackages, I was able to publish voltage values to the GUI without any issues

File: docker/rosbridge/Dockerfile
Comment (MartensCedric) [reply docker/rosbridge/Dockerfile:17]: You might as well commit it my man or make a suggest it through GitHub

Comment (RealWilliamWells) [reply docker/rosbridge/Dockerfile:17]: Yea I guess I'll just commit real quick then

**Review by MartensCedric — CHANGES_REQUESTED** [review]
So I used --build and there seem to have been some good progress :+1: 

When I tried publishing it didnt work. So I knew it was probably because the catkin_make in rospackages didnt happen. Looking deeper in it, it's because it fails

image

**Comment by MartensCedric** [comment]
> So I used --build and there seem to have been some good progress 
> 
> When I tried publishing it didnt work. So I knew it was probably because the catkin_make in rospackages didnt happen. Looking deeper in it, it's because it fails
> 
> image

You most likely need to run rosdep on the `rospackages` directory. Rosdep isn't even installed so that is prob why it doesnt work. 

http://wiki.ros.org/rosdep

**Review by MartensCedric — CHANGES_REQUESTED** [review]
The aliases don't work (not a huge deal)

 But this folder should have way more stuff in it 
image

Where's the basestation folder, the util folder etc?

**Comment by MewtR** [comment]
@MartensCedric 
I added all other folders and sourced `basestation/config/.bash_aliases` as well as `rover/config/.bash_aliases`. See 
5b67962

**Review by MartensCedric — CHANGES_REQUESTED** [review]
C++ doesnt compile. I dmed you so we can meet so I can explain it to you, it should not be very hard

**Comment by MewtR** [comment]
@MartensCedric 
A bit hacky but I got it done:
* See 7707913 for the spaceuser password. (password is also spaceuser)
* See 135549b for the script that installs arduino
* See e8178b1 for the `local_rover.launch` startup

**Comment by MartensCedric** [comment]
image
Having trouble with this build, does it work for you guys @RealWilliamWells @AFknUsername @ginywiny @PeterGhimself ?

File: docker/docker-compose.yml
Comment (ginywiny) [inline docker/docker-compose.yml:1]: I had to downgrade this to version 2.0 to run on Ubuntu18.04. This version was also not compatible with Ubuntu16.04. 
I downloaded the most recent docker-compose and faced this error.

**Comment by RealWilliamWells** [comment]
> image
> Having trouble with this build, does it work for you guys @RealWilliamWells @AFknUsername @ginywiny @PeterGhimself ?

I didn't get to finish building the docker due to a network error and it took forever to get to this point, but it built and installed Teensyduino just fine without any issues.

**Review by ginywiny — COMMENTED** [review]
Hey, after looking at this together during the scrum, I noticed some issues. 
 
1. Tried running on Ubuntu16.04, which did not work due to docker-compose version being inferior. Upon updating to Ubuntu18.04, I ran again, and once again faced version issues. We had to manually downgrade the version from the docker-compose.yml to version 2 for the docker-compose to build. 

2. /mux_select and /task_handler both do not exist. Not sure if this is caused because of the docker-compose version, but I am not sure of this.

3. And to mention this, the package configuration is delayed, since apt-utils is not installed

**Comment by ginywiny** [comment]
> image
> Having trouble with this build, does it work for you guys @RealWilliamWells @AFknUsername @ginywiny @PeterGhimself ?

@MartensCedric I did not face any issues installing Teensyduino.

**Comment by ginywiny** [comment]
I did have some other issues with /mux_select and /task_handler though. Not sure if this is because of running an older version of docker-compose (version 2) as opposed to 3.6, since we had to manually change it for the build to run. 

image
Have you also seen this? @RealWilliamWells @AFknUsername @ginywiny @PeterGhimself

**Comment by MewtR** [comment]
@MartensCedric 
So I figured out why you're getting that error. Basically when you build `robot/rover` directory and then docker-compose some of the files generated by cmake get copied into the container and that messes up the cmake you run in the Dockerfile. The solution was to simply dockerignore the files generated by cmake (See d026e85 & c51dd6f).

Now if you just `git pull` and `docker-compose up --build` you shouldn't have that error anymore.

File: docker/docker-compose.yml
Comment (MewtR) [reply docker/docker-compose.yml:1]: Just to be sure can you tell me what you get when you run `docker-compose --version`

Comment (MartensCedric) [reply docker/docker-compose.yml:1]: for reference im using docker-compose version 1.27.4, build 40524192

**Comment by MartensCedric** [comment]
So the teensyduino error is no more, I had to wait like 10 mins. I thought it was stuck but it wasnt lol. 

Does the cmake build actually work for you @MewtR ? Because it does not for me.
image
Honestly I feel like we should just merge this PR, this is something @steamypotato can probably help investigate if you're stuck on that.

Everything else seems good as the local rover stuff launched.

**Review by MartensCedric — APPROVED** [review]

**Comment by MewtR** [comment]
@MartensCedric
I put the cmake part in the Dockerfile so it should already be built once you enter the container. I haven't tried building it manually like you did (that's what you're trying to do right?)

**Review by MartensCedric — COMMENTED** [review]
> @MartensCedric
> I put the cmake part in the Dockerfile so it should already be built once you enter the container. I haven't tried building it manually like you did (that's what you're trying to do right?)

Yes. Having it build when the container is create is not useful. Having it built inside the container on demand can allow someone to develop C++ code inside of docker so it has more value

**Comment by MartensCedric** [comment]
> > @MartensCedric
> > I put the cmake part in the Dockerfile so it should already be built once you enter the container. I haven't tried building it manually like you did (that's what you're trying to do right?)
> 
> Yes. Having it build when the container is create is not useful. Having it built inside the container on demand can allow someone to develop C++ code inside of docker so it has more value

But I approved the PR because its kind of getting out of scope I think, we may as well open a new issue. Im just waiting for a second approver to merge @PeterGhimself @AFknUsername @RealWilliamWells

**Comment by MewtR** [comment]
@MartensCedric 
Alright that's fine with me but we should discuss the workflow maybe. Because in my head it was more like all the development/editing is done on the host (outside of the containers) and then when it's time to run you do `docker-compse up --build`. Although there are a few more changes that need to be made so that this sort of workflow doesn't get annoying because of how long it takes to build the containers.

**Comment by MartensCedric** [comment]
> @MartensCedric
> Alright that's fine with me but we should discuss the workflow maybe. Because in my head it was more like all the development/editing is done on the host (outside of the containers) and then when it's time to run you do `docker-compse up --build`. Although there are a few more changes that need to be made so that this sort of workflow doesn't get annoying because of how long it takes to build the containers.

ok then, this makes more sense. I always thought it was a pain to work directly in docker. Makes more sense your approach. This is why then we should approve and merge this since it works

File: docker/docker-compose.yml
Comment (ginywiny) [reply docker/docker-compose.yml:1]: @MewtR docker-compose version 1.27.4, build unknown is what I have setup

Comment (ginywiny) [reply docker/docker-compose.yml:1]: Turns out using pip3 install docker-compose causes an unknown build to be set

**Review by ginywiny — COMMENTED** [review]
**Docker-version problem fixed.**
After looking into my `docker-compose --version`, it turns out that the build is "unknown". I have reinstalled using docker's official instructions, and it turns out that I would always end up with this issue when installing with `pip3 install docker-compose`. 

Upon speaking to @MartensCedric, I just decided to run a basic apt install and ended up with the same version `docker-compose version 1.27.4, build 40524192`.

Using the apt installed docker-compose, I was able to properly run docker-compose up --build, for version 3.6 in the docker-compose.yml. 

**Noticed during build**
While running this build, I saw that there were some deprecated declarations.
image

Is this normal? I am running the latest version of the PR. Has anybody else faced this issue? @MartensCedric  @RealWilliamWells  @AFknUsername

**Comment by MartensCedric** [comment]
> **Docker-version problem fixed.**
> After looking into my `docker-compose --version`, it turns out that the build is "unknown". I have reinstalled using docker's official instructions, and it turns out that I would always end up with this issue when installing with `pip3 install docker-compose`.
> 
> Upon speaking to @MartensCedric, I just decided to run a basic apt install and ended up with the same version `docker-compose version 1.27.4, build 40524192`.
> 
> Using the apt installed docker-compose, I was able to properly run docker-compose up --build for version 3.6 in the docker-compose.yml.
> 
> **Noticed during build**
> While running this build, I saw that there were some deprecated declarations.
> image
> 
> Is this normal? I am running the latest version of the PR. Has anybody else faced this issue? @MartensCedric @RealWilliamWells @AFknUsername

this is a build of a 3rd party software that we did not write so whatever warnings they have is not part of our project

**Review by RealWilliamWells — APPROVED** [review]
Just tested and everything works as intended. Good job 👌

**Comment by AFknUsername** [comment]
> > **Docker-version problem fixed.**
> > After looking into my `docker-compose --version`, it turns out that the build is "unknown". I have reinstalled using docker's official instructions, and it turns out that I would always end up with this issue when installing with `pip3 install docker-compose`.
> > 
> > Upon speaking to @MartensCedric, I just decided to run a basic apt install and ended up with the same version `docker-compose version 1.27.4, build 40524192`.
> > 
> > Using the apt installed docker-compose, I was able to properly run docker-compose up --build for version 3.6 in the docker-compose.yml.
> > 
> > **Noticed during build**
> > While running this build, I saw that there were some deprecated declarations.
> > image
> > 
> > Is this normal? I am running the latest version of the PR. Has anybody else faced this issue? @MartensCedric @RealWilliamWells @AFknUsername
> 
> this is a build of a 3rd party software that we did not write so whatever warnings they have is not part of our project

Yeah, those errors are ok. Hopefully the package will get updated eventually to get rid of them.

**Pull request merged**
